docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,280 @@
+for each minor release.
+
+#### Installing an Older Version of a z-stream
+
+As a user of assisted-installer UI, I should have the option to browse the
+different available versions that are fetched from upgrade channels.
+
+#### Configuring the Dynamically-loaded Versions of a Stream
+
+As a system administrator for the assisted-installer, I should have the option
+to configure both dynamically-loaded and statically-loaded OpenShift version.
+To do that, I can provide the following environment variables for example:
+```
+RELEASE_IMAGES='[
+  {
+    "openshift_version": "4.12-multi",
+    "cpu_architecture": "multi",
+    "cpu_architectures": [
+        "x86_64",
+        "arm64",
+        "ppc64le",
+        "s390x"
+    ],
+    "url": "quay.io/openshift-release-dev/ocp-release:4.12.0-rc.4-multi",
+    "version": "4.12.0-rc.4"
+  }
+]'
+
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='[
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+        {
+          "cpu_architecture": "x86_64",
+          "channels": ["stable"]
+        }
+    ]
+  },
+  {
+    "openshift_version": "4.11",
+    "upgrade_channels": [
+      {
+        "cpu_architecture": "x86_64",
+        "channels": ["candidate", "stable"]
+      },
+      {
+        "cpu_architecture": "arm64",
+        "channels": ["candidate", "stable"]
+      }
+    ]
+  }
+]'
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+Mostly conveyed in ``Risks and Mitigations``.
+
+### Risks and Mitigations
+
+We have recently seen occasions where the disk space is being exhausted due to
+the assisted-service caching a lot of different openshift installer binaries.
+It is important to make sure we don't make this problem more severe, as this
+enhancement introduces much more versions which may (or may not) cause use of
+them.
+A possible solution here is implementing the cache as a LRU-like cache, making
+sure we evict the old entries when we reach a certain capacity.
+
+We should make sure we don't consume too much memory when fetching releases.
+It's especially critical when the user queries for the different versions, but
+might happen in other cases too.
+
+We need to think of the retention policy of versions. We don't expect to have
+versions removed from the upgrade-channels API, but just in case we should only
+add versions and not mirror exactly the given result of those API responses.
+As opposed to that, when a minor release gets EOL (end-of-life) and we remove it
+from the configuration, the service should remove all the stale versions from
+the database.
+
+## Design Details [optional]
+
+If an enhancement is complex enough, design details should be included. When not
+included, reviewers reserve the right to ask for this section to be filled in to
+enable more thoughtful discussion about the enhancement and it's impact.
+
+### Open Questions
+
+* Is the database usage a real necessity? The idea is having the periodic thread
```

Comment:
Exactly. If we're gonna fetch periodically anyway, maybe we can just cache the result for a while.

Reply:
will it work well when being done from different threads?

Reply:
I think [go-cache](https://github.com/patrickmn/go-cache) we usually use is thread safe, but we can also try just using [gocron](https://github.com/go-co-op/gocron) instead of our monitor.

Reply:
I don't think we need polling here. We should be able to look up new versions when the user asks for one we don't know about.

Reply:
We already have a cache for versions in `internal/versions/versions.go` that should just continue to be the cache.

Reply:
> I don't think we need polling here. We should be able to look up new versions when the user asks for one we don't know about.

@carbonin the suggested U/X is to provide the user with available versions. It means deciding on that list on the assisted-service level
do you mean having the UI gather all the versions and let the client-side set a version for a cluster? (I just want to make sure I understand your suggestion)

Reply:
+1 on that. We can also just invoke the query on cache expire. I.e. if the value in cache is [stale](https://github.com/patrickmn/go-cache/blob/master/cache.go#L128), fetch the list from api.openshift

Reply:
I'm on the fence here, because it is much simpler but has a major disadvantage. If this logic happens in the same thread of the http server, some requests can stall with their response for too long

I need to benchmark it a bit to understand if it's really substantial on the nominal case
Either way remember it's a dependency on an external API that can also be down / have an outage / not perform well. We wouldn't want to wait too long for its responses to return

Reply:
We use a similar mechanism when calling AMS, so I think it should be safe enough. E.g. https://github.com/openshift/assisted-service/blob/master/pkg/auth/rhsso_authz_handler.go
There's also the fallback to the current release images list, right? So in case of an error/timeout we'd just return the fallback list.

Reply:
Indeed we should be able to configure a client-side timeout.
We will want to ensure we protect the in-memory release image list with a locking mechanism though.

I had considered this when doing the versions refactor, but we'll definitely need it now.

Reply:
Ack. I'll change to in-memory implementation and mention caveats of this solution in the mitigations & risks section